### PR TITLE
Upgrade: No Code Metadata Modeling

### PR DESCRIPTION
**Pull request by jjoyce0510**
**No Code Metadata**

This PR introduces a major refactor that permits extension of DataHub’s metadata model without writing any imperative code. 

It was the subject of the final tech talk at the May Community All Hands:

- [Slides](https://docs.google.com/presentation/d/1_F2xy2agb7M4IFhAKy_ORte0w-7hBRfTCypxwWhr2kg/edit?usp=sharing)
- [Recording](https://www.youtube.com/watch?v=mjKjjtm8GfM)

 You can find an abbreviated recap below.  

**The Problem**

Adding an entity to GMS requires too many changes. 
- \> 25 files: Snapshot Models, Document models, Relationship models, Entity models, Rest.li Resource models, Rest.li Resource + Subresource impls, Graph Builders, Index Builders, index settings + mappings, + much more. 
- \> Takes 1-2 weeks, not counting PR back and forth
         - Adding Dashboards and Charts Backend: https://github.com/linkedin/datahub/pull/1884/files 
         - Adding DataJobs and DataFlows: https://github.com/linkedin/datahub/pull/2197/files 
         - Adding business glossary term + : https://github.com/linkedin/datahub/pull/2228/files 
         - Adding ML model and feature: https://github.com/linkedin/datahub/pull/1896/files 
         - Extremely difficult to document the process with high quality, due to sheer number of steps 

**The Goal**
Adding / Extending a GMS entity takes <= 15 minutes 

**The Solution (Changes)**

1. Introduce custom PDL model annotations for declaring
       - Entity and Aspect Models 
       - Indexable fields
       - Foreign key Relationships
       - Generalize Rest.li Resources by introducing generic Entity, Relationship Resources that allow you to fetch, browse, search, and find relationships for any GMS  entity or relationship 
2. Generalize the aspect-specific local (key-value) dao, search daos, browse daos, and graph daos such that one instance can be used to write any arbitrary aspect (relax type constraints by moving away from heavy use of java generics) to new Service abstraction
3. Remove hardcoded logic to compute browse path, instead provide browse path dynamically via newly introduced aspect
4. Remove the requirement to define an Urn typeref + hand-code an Urn.java class, replacing with use of generic Urns + PDL “key” aspects whose fields map to urn parts. 
5. Update MCE processor to leverage generic Entity Resource (ref. Step 1) 
6. Update MAE processor to populate Graph store and Search index driven by model annotations (ref. Step 3) 
7. Migrate GraphQL server to leverage new generic Rest.li resources (ref. Step 1)
8. Migrate ingestion framework to leverage generic Rest.li resources (ref. Step 1) 
9. Introduced “datahub-upgrade” container for performing upgrades between DataHub versions. Two upgrades are implemented to support this PR:
          - NoCodeDataMigration: Migrates a running instance of DataHub from legacy storage to new storage models. 
          - NoCodeDataMigrationCleanup: Removes legacy data. Should only be performed once you have verified No Code upgrade has completed successfully. 

**The Impact**

With these changes, we’ve reduced the time to add or extend an entity from 2 weeks+ to 15 minutes. Additionally, we’ve made over 270 files in DataHub unnecessary. These will be removed in a fast followup. 

**The Migration**

As part of these changes, we’ve decoupled the persistence layer from Java + Rest.li details. Previously, the aspect name was derived by taking the fully qualified classname of the corresponding to an auto-generated Rest.li Record Template Java class. 

Now, we use the more explicit aspect model annotations (@Aspect)  to obtain common aspect names (eg. instead of “com.linkedin.metadata.common.Ownership” we use a more readable “ownership”). 

Same goes for the graph store, where edge and node data included fully-qualified classnames of entity classes and relationship classes. 

This has impacts on customers who have already deployed DataHub and for whom restoring the data from scratch is not an option.

Thus, we’re including an upgrade CLI + containers to allow these customers to seamlessly upgrade their persistence layer from the previous version (which we are denoting “v1”) to the new storage version (v2). 

The upgrade logic is as follows:

1. Creates the new v2 SQL table if it does not exist 
2. Reads from the v1 source of truth SQL table in batches
3. Performs the necessary reverse mapping from aspect classname -> new aspect common name 
4. Writes batches back into a new v2 SQL table
5. Forces emission of MAE when writing aspects where version = latest to ensure indices are rebuilt properly 

The batch size and delay between batch writes can be configured via environment variables provided to the script. 

As a pre-requisite to executing the script

- The latest version of GMS should be pulled and deployed. 
- The latest version of MAE consumer should be pulled and deployed. 

Upon migrating successfully, customers may run a provided cleanup script to cleanse your stores and indexes of old data. 

For more information, see the Migration Guide documentation. 

**The Documentation**

The following documentation is included within this PR:

- [Technical Overview](https://github.com/linkedin/datahub/blob/5c6dc25e734a23c7918421dd0d39a034e8dc40af/docs/advanced/no-code-modeling.md)
- [The DataHub Metadata Model  ](https://github.com/linkedin/datahub/blob/5c6dc25e734a23c7918421dd0d39a034e8dc40af/docs/modeling/metadata-model.md)
- [Extending the Metadata Model](https://github.com/linkedin/datahub/blob/5c6dc25e734a23c7918421dd0d39a034e8dc40af/docs/modeling/extending-the-metadata-model.md)
- [No Code Upgrade Guide ](https://github.com/linkedin/datahub/blob/5c6dc25e734a23c7918421dd0d39a034e8dc40af/docs/advanced/no-code-upgrade.md)

**The Future**

In rapid follow ups, we will begin to clean the DataHub repository of the 271 files that have been made unnecessary due to these changes. Stay tuned! 

**Support**

Do not hesitate to reach out to the Acryl team for assistance adopting this version of DataHub:

- @jjoyce0510 
- @dexter-mh-lee 
- @gabe-lyons 
- @shirshanka 

## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,6 +8,9 @@
 *.iws
 *.ipr
 
+*gma-create-all.sql
```
Comment (shirshanka) [inline .gitignore:11]: why are these important?

Comment (jjoyce0510) [reply .gitignore:11]: These are auto-generated by Ebean test framework, thus we want to omit them: https://ebean.io/docs/ddl-generation/

**Comment by nagarjunakanamarlapudi** [comment]
@jjoyce0510  Do you think it is better to separate the RFC PR and implementations PR ? Hard to follow what changes are proposed.

**Review by shirshanka — APPROVED** [review]
LGTM! 
Let's go!

**Comment by jjoyce0510** [comment]
@nagarjunakanamarlapudi Let's chat more. I can try to clarify anything that is unclear. For now, here's the tech overview doc: https://datahubproject.io/docs/advanced/no-code-modeling/

Excited to get your feedback and improve what we've built!
